### PR TITLE
bpo-34741: Get rid of tp_getattro and tp_setattro in pyexpat.xmlparser.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
Use tp_members and tp_getset instead.

https://bugs.python.org/issue34741

### CODE REVIEW

File: Modules/pyexpat.c [hunk 1]
```diff
@@ -1234,221 +1170,151 @@ xmlparse_dealloc(xmlparseobject *self)
...
-
-    if (handlernum != -1) {
-        PyObject *result = self->handlers[handlernum];
-        if (result == NULL)
-            result = Py_None;
-        Py_INCREF(result);
-        return result;
+    int handlernum = hi - handler_info;
+    if (v == NULL) {
+        PyErr_SetString(PyExc_RuntimeError, "Cannot delete attribute");
```
Comment (scoder) [inline Modules/pyexpat.c]: Why not `AttributeError` ?

File: Modules/pyexpat.c [hunk 2]
```diff
@@ -1234,221 +1170,151 @@ xmlparse_dealloc(xmlparseobject *self)
...
-            v = NULL;
-        }
-        else if (v != NULL) {
-            Py_INCREF(v);
-            c_handler = handler_info[handlernum].handler;
-        }
-        Py_XSETREF(self->handlers[handlernum], v);
-        handler_info[handlernum].setter(self->itself, c_handler);
-        return 1;
+#define INT_GETTER(name) \
```
Comment (scoder) [inline Modules/pyexpat.c]: Maybe `#undef` it after use, to limit the scope of the macro?

File: Modules/pyexpat.c [hunk 3]
```diff
@@ -1234,221 +1170,151 @@ xmlparse_dealloc(xmlparseobject *self)
...
 {
-    /* Set attribute 'name' to value 'v'. v==NULL means delete */
-    if (!PyUnicode_Check(name)) {
-        PyErr_Format(PyExc_TypeError,
-                     "attribute name must be string, not '%.200s'",
-                     name->ob_type->tp_name);
-        return -1;
-    }
     if (v == NULL) {
         PyErr_SetString(PyExc_RuntimeError, "Cannot delete attribute");
```
Comment (scoder) [inline Modules/pyexpat.c:1255]: Same here, why not `AttributeError` ?

File: Modules/pyexpat.c [hunk 4]
```diff
@@ -1635,6 +1582,32 @@ static struct PyModuleDef pyexpatmodule = {
         NULL
 };
 
+static int init_handler_descrs(void)
```
Comment (scoder) [inline Modules/pyexpat.c:1590]: This function looks _really_ low-level. Isn't there a more "official" API way to do this?
And why is this called after `PyType_Ready()` and does not invalidate the type/method cache?

**Review by scoder — COMMENTED** [review]
Left a couple of comments. It generally looks like a nice cleanup. Thanks Serhiy!

File: Modules/pyexpat.c (hunk 1 above)
Comment (serhiy-storchaka) [reply Modules/pyexpat.c]: Because the current code raises RuntimeError.

File: Modules/pyexpat.c (hunk 2 above)
Comment (serhiy-storchaka) [reply Modules/pyexpat.c]: Good point.

File: Modules/pyexpat.c (hunk 3 above)
Comment (serhiy-storchaka) [reply Modules/pyexpat.c:1255]: This code was not changed.

File: Modules/pyexpat.c (hunk 4 above)
Comment (serhiy-storchaka) [reply Modules/pyexpat.c:1590]: I think this is an official way. Only public API is used.

Calling `PyType_Ready()` is needed for initializing `tp_dict`.

At that time the type doesn't have valid method cache.

**Pull request merged**